File: backend/app/schemas/task_report.py
import uuid

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class TaskReportPeriod(BaseModel):
    model_config = ConfigDict(extra="forbid")

    planned_from: date | None
    planned_to: date | None


class TaskOutcomeMetrics(BaseModel):
    model_config = ConfigDict(extra="forbid")

    completed_count: int = Field(ge=0)
    not_completed_count: int = Field(ge=0)
    terminal_count: int = Field(ge=0)
    completion_rate: Decimal | None = Field(default=None, ge=0, le=100)


class TaskMasterTaskReportRow(TaskOutcomeMetrics):
    master_task_id: uuid.UUID
    master_task_name: str
    category_id: uuid.UUID
    category_name: str


class TaskReportResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    period: TaskReportPeriod
    summary: TaskOutcomeMetrics
    by_master_task: list[TaskMasterTaskReportRow]

File: frontend/src/types/taskReport.ts
export interface TaskReportParams { planned_from?: string; planned_to?: string; master_task_id?: string; category_id?: string }
export interface TaskOutcomeMetrics { completed_count:number; not_completed_count:number; terminal_count:number; completion_rate:string|number|null }
export interface TaskReportRow extends TaskOutcomeMetrics { master_task_id:string; master_task_name:string; category_id:string; category_name:string }
export interface TaskReportResponse { period:{planned_from:string|null;planned_to:string|null}; summary:TaskOutcomeMetrics; by_master_task:TaskReportRow[] }
